### PR TITLE
Issue #157 custom mail templates

### LINKED ISSUE #157: Custom mail templates source
### Checklist

*   [x] I read [Contribution Guidelines](https://github.com/apragacz/django-rest-registration/blob/master/CONTRIBUTING.md#issues)
*   [x] I searched [the documentation](https://django-rest-registration.readthedocs.io/) to ensure that the requested feature is not already implemented and described
*   [x] I searched existing issues before opening this one

### Is your feature request related to a problem? Please describe.
In some of my applications, email templates sent to customers are stored in a database (as HTML). Right now this plugin requires storing template in the project repository and configure a path to its location. I couldn't find a way to override this behaviour without modifying the plugin's source.

### Describe the solution you'd like
To provide the ability to supply own HTML template there could be a function that converts arbitrary `template_config_data` straight into an object containing `subject`, `text_body`, and `html_body`. Then there could be a settings param `VERIFICATION_TEMPLATE_BUILDER` that by default will point to that function.

### Describe alternatives you've considered
Similar functionality could be added after migrating to class-based views by creating some `BaseTemplateBuilder` class.

I'll be more than happy to help with this feature as this plugin is exactly what I need for my project except the ability to supply email templates from a database.

### PR DESCRIPTION
**Pull request by rarvolt**
This PR adds the ability to customize HTML generation using other sources than template files.
Resolves #157

### CODE REVIEW

File: rest_registration/utils/verification.py [hunk 1]
```diff
@@ -49,6 +55,31 @@ def build_default_template_context(
     return context
 
 
+def build_default_template(
+    user: 'AbstractBaseUser',
+    template_config_data: Dict[str, Any],
```
Comment (apragacz) [inline rest_registration/utils/verification.py:60]: I think that function should be more general. In my opinion it should receive:

*   ``user`` - the user which is to be notified.
*   ``notification_type`` - value of ``rest_registration.notifications.enums.NotificationType`` enum.
*   ``notification_method`` - value of ``rest_registration.notifications.enums.NotificationMethod``; will be hardcoded for `EMAIL` for now in the caller function
*   ``template_config_data`` - about this one I'm unsure, but as it depends on `request` which will be unavailable here so it should probably be included as well.

Of course, docs for the setting need to be updated as well.

**Review by apragacz — COMMENTED** [review]
@rarvolt thanks for providing PR with proposed solution!

Overall the change looks good, but I have some concerns related to naming and args passed to the rendering function (see specific comments).

File: rest_registration/utils/email.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    'html_body_template_name',
+    'text_body_processor',
+))
+
+
+def get_html_to_text_converter() -> Callable[[str], str]:
+    return registration_settings.VERIFICATION_EMAIL_HTML_TO_TEXT_CONVERTER  # noqa: E501
+
+
+def parse_template_config(template_config_data: Dict[str, Any]) -> EmailTemplateConfig:
```
[the remark this answers is not in the document]
Comment (rarvolt) [reply rest_registration/utils/email.py:24]: Yes, exactly.

File: rest_registration/settings_fields.py [hunk 3]
```diff
@@ -374,6 +374,30 @@ def __new__(
             of the function.
             """),
     ),
+    Field(
+        'VERIFICATION_TEMPLATE_BUILDER',
```
[the remark this answers is not in the document]
Comment (rarvolt) [reply rest_registration/settings_fields.py:378]: I wasn't sure about `BUILDER`. `RENDERER` sounds good.

File: rest_registration/utils/verification.py [hunk 4]
```diff
@@ -49,6 +55,31 @@ def build_default_template_context(
     return context
 
 
+def build_default_template(
```
[the remark this answers is not in the document]
Comment (rarvolt) [reply rest_registration/utils/verification.py:58]: Sounds good.

File: rest_registration/utils/verification.py [hunk 5]
```diff
@@ -1,18 +1,24 @@
...
 from rest_registration.notifications.enums import NotificationMethod, NotificationType
 from rest_registration.settings import registration_settings
+from rest_registration.utils.email import parse_template_config
 from rest_registration.utils.signers import URLParamsSigner
 
 if TYPE_CHECKING:
     from django.contrib.auth.base_user import AbstractBaseUser
 
 
+TemplateData = namedtuple('TemplateData', 'subject,text_body,html_body')
```
[the remark this answers is not in the document]
Comment (rarvolt) [reply rest_registration/utils/verification.py:19]: Let's go for `EmailTemplateRenderResult`

File: rest_registration/utils/verification.py (hunk 1 above)
Comment (rarvolt) [reply rest_registration/utils/verification.py:60]: This function receives `template_config_data` which is a generic dict generated in `_get_email_template_config_data` using function set with `VERIFICATION_TEMPLATES_SELECTOR`.
`template_selector` already receives `request`, `user`, `notification_method` and `notification_type` (here hardcoded). I think it is redundant to pass the same data here. If one really wanted to access eg. `request`, he could override selector function to pass additional data into the `template_config_data`.

Comment (apragacz) [reply rest_registration/utils/verification.py:60]: That argument makes sense. But in that case, my question is why we need to provide `user` here? Information about the user can be encoded in `template_config` and/or in `context` (as both can be controlled via `VERIFICATION_TEMPLATES_SELECTOR` and `VERIFICATION_TEMPLATE_CONTEXT_BUILDER`), so I'm not sure if that one is not redundant as well? I don't see `user` being used below.

Comment (rarvolt) [reply rest_registration/utils/verification.py:60]: I tried to follow a pattern where most functions got a `user` argument, but you are right:
- `VERIFICATION_TEMPLATES_SELECTOR` can select/generate correct template data based on the `user`
- `VERIFICATION_TEMPLATE_CONTEXT_BUILDER` can create correct context with `user` details
I'll remove this argument.

**Comment by apragacz** [comment]
@rarvolt I added a commit fixing the remaining isort issue. Do you plan to add anything there? If not, I can merge it and release new version.

**Review by apragacz — APPROVED** [review]

**Comment by rarvolt** [comment]
@apragacz I've been using new functionality in a testing version of my app. It works well so you can go ahead and merge this.

**Comment by apragacz** [comment]
@rarvolt why the PR was closed? Can I re-open it and merge it? Or is it something I should be aware of?

**Comment by rarvolt** [comment]
@apragacz sorry about that, I don't do a lot of PRs in GitHub so I didn't know how that mechanism works ;)

**Comment by apragacz** [comment]
Sure, no problem :)

**Pull request merged**